File: python/src/datagrout/conduit/registration.py
"""Substrate identity registration with the DataGrout CA.

The registration flow mirrors the Rust SDK exactly (Rust is the reference spec):

1. :func:`generate_keypair` — generate an ECDSA P-256 keypair locally.
   Returns a :class:`~datagrout.conduit.ConduitIdentity` holding the private
   key and a **temporary self-signed certificate** (placeholder only).
   The private key never leaves the client.

2. :func:`register_identity` — send only the public key to DataGrout.
   DG signs the cert and returns a DG-CA-signed certificate.
   The returned identity replaces the placeholder cert with the real one.

3. :func:`save_identity` — persist the returned identity to ``~/.conduit/``
   for auto-discovery by :meth:`~datagrout.conduit.ConduitIdentity.try_default`.

4. :func:`rotate_identity` — on renewal (cert near expiry), present the
   *existing* client cert over mTLS (no API key needed) to get a new one.

CA cert refresh
---------------
:func:`fetch_dg_ca_cert` fetches the current DataGrout CA certificate from
``https://ca.datagrout.ai/ca.pem``. This uses the system trust store, so
there is no circularity. Call :func:`refresh_ca_cert` at startup to pick up
CA rotations without re-registering.
"""

from __future__ import annotations

import datetime
from dataclasses import dataclass
from pathlib import Path
from typing import Optional, Tuple

import httpx

from .identity import ConduitIdentity

#: Canonical URL for the DataGrout CA certificate.
DG_CA_URL = "https://ca.datagrout.ai/ca.pem"

#: Default local identity directory.
DEFAULT_IDENTITY_DIR = Path.home() / ".conduit"


# ─── Receipt / meta types ─────────────────────────────────────────────────────


@dataclass
class Byok:
    """BYOK discount details embedded in a receipt."""
    enabled: bool = False
    discount_applied: float = 0.0
    discount_rate: float = 0.0


@dataclass
class Receipt:
    """Cost receipt attached to every DG tool-call result under ``_meta.receipt``.

    Use :func:`extract_meta` to pull this out of any tool-call result dict.
    """
    receipt_id: str
    timestamp: str
    estimated_credits: float
    actual_credits: float
    net_credits: float
    savings: float
    savings_bonus: float
    breakdown: dict
    byok: Byok
    transaction_id: Optional[str] = None
    balance_before: Optional[float] = None
    balance_after: Optional[float] = None


@dataclass
class CreditEstimate:
    """Pre-execution credit estimate under ``_meta.credit_estimate``."""
    estimated_total: float
    actual_total: float
    net_total: float
    breakdown: dict


@dataclass
class ToolMeta:
    """The ``_meta`` block DataGrout appends to every tool-call result.

    Example::

        result = await client.call_tool("salesforce@v1/get_lead@v1", {...})
        meta = extract_meta(result)
        if meta:
            print(f"Charged {meta.receipt.net_credits} credits")
    """
    receipt: Receipt
    credit_estimate: Optional[CreditEstimate] = None


def extract_meta(result: dict) -> Optional[ToolMeta]:
    """Extract the DataGrout metadata block from a tool-call result dict.

    Checks ``_datagrout`` first (current format), then falls back to ``_meta``
    for backward compatibility with older gateway responses.

    Returns ``None`` when the result contains neither key (e.g. upstream
    servers not routed through the DG gateway).
    """
    raw_meta = None
    if isinstance(result, dict):
        raw_meta = result.get("_datagrout") or result.get("_meta")
    if not raw_meta:
        return None

    raw_receipt = raw_meta.get("receipt", {})
    receipt = Receipt(
        receipt_id=raw_receipt.get("receipt_id", ""),
        timestamp=raw_receipt.get("timestamp", ""),
        estimated_credits=raw_receipt.get("estimated_credits", 0.0),
        actual_credits=raw_receipt.get("actual_credits", 0.0),
        net_credits=raw_receipt.get("net_credits", 0.0),
        savings=raw_receipt.get("savings", 0.0),
        savings_bonus=raw_receipt.get("savings_bonus", 0.0),
        breakdown=raw_receipt.get("breakdown", {}),
        byok=Byok(**raw_receipt.get("byok", {})),
        transaction_id=raw_receipt.get("transaction_id"),
        balance_before=raw_receipt.get("balance_before"),
        balance_after=raw_receipt.get("balance_after"),
    )

    raw_est = raw_meta.get("credit_estimate")
    credit_estimate = (
        CreditEstimate(
            estimated_total=raw_est.get("estimated_total", 0.0),
            actual_total=raw_est.get("actual_total", 0.0),
            net_total=raw_est.get("net_total", 0.0),
            breakdown=raw_est.get("breakdown", {}),
        )
        if raw_est
        else None
    )

    return ToolMeta(receipt=receipt, credit_estimate=credit_estimate)


# ─── Registration response ─────────────────────────────────────────────────────


@dataclass
class RegistrationResponse:
    """Response body from ``POST /api/v1/substrate/identity/register`` or ``/rotate``."""
    id: str
    cert_pem: str
    fingerprint: str
    name: str
    registered_at: str
    ca_cert_pem: Optional[str] = None
    valid_until: Optional[str] = None


@dataclass
class SavedIdentityPaths:
    """Paths written by :func:`save_identity`."""
    cert_path: Path
    key_path: Path
    ca_path: Optional[Path] = None


# ─── CA cert fetching ─────────────────────────────────────────────────────────


async def fetch_dg_ca_cert(url: str = DG_CA_URL) -> str:
    """Fetch the current DataGrout CA certificate from ``ca.datagrout.ai``.

    Uses the system trust store for HTTPS — no circularity with the DG CA.

    Returns:
        PEM-encoded CA certificate string.
    """
    async with httpx.AsyncClient() as client:
        resp = await client.get(url, headers={"Accept": "application/x-pem-file, text/plain"})
        resp.raise_for_status()
        pem = resp.text

    if "-----BEGIN CERTIFICATE-----" not in pem:
        raise ValueError(f"Response from {url} does not look like a PEM certificate")

    return pem


async def refresh_ca_cert(
    identity_dir: Path = DEFAULT_IDENTITY_DIR,
    url: str = DG_CA_URL,
) -> Path:
    """Refresh the locally-cached DG CA certificate.

    Fetches the current CA cert and writes it to ``{identity_dir}/ca.pem``.
    Call at application startup to pick up CA rotations transparently.

    Returns:
        Path to the written ``ca.pem`` file.
    """
    pem = await fetch_dg_ca_cert(url)
    identity_dir.mkdir(parents=True, exist_ok=True)
    ca_path = identity_dir / "ca.pem"
    ca_path.write_text(pem)
    return ca_path


# ─── Keypair generation ───────────────────────────────────────────────────────


def generate_keypair(name: str) -> ConduitIdentity:
    """Generate an ECDSA P-256 keypair for Substrate identity registration.

    Returns a :class:`~datagrout.conduit.ConduitIdentity` holding the private
    key and a **temporary self-signed certificate** (placeholder only).  Pass
    the returned object to :func:`register_identity` to exchange the placeholder
    for a DG-CA-signed certificate.

    The private key is generated entirely locally and never transmitted.

    Args:
        name: Human-readable label embedded in the self-signed cert CN.

    Returns:
        A :class:`ConduitIdentity` with a placeholder cert; call
        :func:`register_identity` to get the real DG-signed identity.
    """
    from cryptography import x509
    from cryptography.hazmat.primitives import hashes, serialization
    from cryptography.hazmat.primitives.asymmetric import ec
    from cryptography.x509.oid import NameOID

    private_key = ec.generate_private_key(ec.SECP256R1())

    subject = x509.Name([x509.NameAttribute(NameOID.COMMON_NAME, name)])
    now = datetime.datetime.now(datetime.timezone.utc)

    cert = (
        x509.CertificateBuilder()
        .subject_name(subject)
        .issuer_name(subject)
        .public_key(private_key.public_key())
        .serial_number(x509.random_serial_number())
        .not_valid_before(now)
        .not_valid_after(now + datetime.timedelta(days=1))
        .sign(private_key, hashes.SHA256())
    )

    cert_pem = cert.public_bytes(serialization.Encoding.PEM).decode()
    key_pem = private_key.private_bytes(
        serialization.Encoding.PEM,
        serialization.PrivateFormat.PKCS8,
        serialization.NoEncryption(),
    ).decode()

    return ConduitIdentity(cert_pem=cert_pem, key_pem=key_pem)


# ─── Registration ─────────────────────────────────────────────────────────────


async def register_identity(
    keypair: ConduitIdentity,
    *,
    endpoint: str,
    api_key: str,
    name: str,
    ca_url: str = DG_CA_URL,
) -> Tuple[ConduitIdentity, RegistrationResponse]:
    """Register a Substrate keypair with the DataGrout CA.

    Sends only the public key to DataGrout and receives back a DG-CA-signed
    certificate.  The private key in *keypair* is reused — it never leaves
    the client.

    Args:
        keypair: A keypair generated by :func:`generate_keypair`.
        endpoint: Base URL for the substrate identity API,
                  e.g. ``https://app.datagrout.ai/api/v1/substrate/identity``.
        api_key: Arbiter API key for bootstrap authentication.
        name: Human-readable label for this Substrate instance.
        ca_url: URL to fetch the DG CA cert from (default: :data:`DG_CA_URL`).

    Returns:
        Tuple of ``(registered ConduitIdentity, RegistrationResponse)``.
        The identity contains the same private key as *keypair* but with the
        DG-signed certificate replacing the placeholder.
    """
    from cryptography.hazmat.primitives import serialization
    from cryptography.hazmat.primitives.asymmetric.ec import EllipticCurvePrivateKey

    # Load the private key from the keypair PEM so we can export the public key.
    from cryptography.hazmat.primitives.serialization import load_pem_private_key
    private_key = load_pem_private_key(keypair.key_pem.encode(), password=None)
    public_key_pem = private_key.public_key().public_bytes(
        serialization.Encoding.PEM,
        serialization.PublicFormat.SubjectPublicKeyInfo,
    ).decode()

    url = endpoint.rstrip("/") + "/register"
    async with httpx.AsyncClient() as client:
        resp = await client.post(
            url,
            headers={"Authorization": f"Bearer {api_key}"},
            json={"public_key_pem": public_key_pem, "name": name},
        )
        resp.raise_for_status()
        body = resp.json()

    ca_pem: Optional[str] = body.get("ca_cert_pem")
    if not ca_pem:
        try:
            ca_pem = await fetch_dg_ca_cert(ca_url)
        except Exception:
            ca_pem = None

    registered = ConduitIdentity(
        cert_pem=body["cert_pem"],
        key_pem=keypair.key_pem,
        ca_pem=ca_pem,
    )
    response = RegistrationResponse(
        id=body["id"],
        cert_pem=body["cert_pem"],
        ca_cert_pem=ca_pem,
        fingerprint=body["fingerprint"],
        name=body.get("name", name),
        registered_at=body.get("registered_at", ""),
        valid_until=body.get("valid_until"),
    )
    return registered, response


# ─── Rotation ─────────────────────────────────────────────────────────────────


async def rotate_identity(
    current_identity: ConduitIdentity,
    *,
    endpoint: str,
    name: str,
    ca_url: str = DG_CA_URL,
) -> Tuple[ConduitIdentity, RegistrationResponse]:
    """Rotate the Substrate identity by presenting the current cert over mTLS.

    Generates a new keypair and sends the public key to the ``/rotate``
    endpoint, authenticated with the *existing* client certificate over mTLS
    — no API key needed.

    Args:
        current_identity: The currently-active identity (used for mTLS).
        endpoint: Base URL for the substrate identity API.
        name: Human-readable label for the renewed identity.
        ca_url: URL to fetch the DG CA cert from.

    Returns:
        Tuple of ``(new ConduitIdentity, RegistrationResponse)``.
    """
    new_keypair = generate_keypair(name)

    from cryptography.hazmat.primitives import serialization
    from cryptography.hazmat.primitives.serialization import load_pem_private_key
    private_key = load_pem_private_key(new_keypair.key_pem.encode(), password=None)
    public_key_pem = private_key.public_key().public_bytes(
        serialization.Encoding.PEM,
        serialization.PublicFormat.SubjectPublicKeyInfo,
    ).decode()

    url = endpoint.rstrip("/") + "/rotate"
    ssl_context = current_identity.build_ssl_context()

    async with httpx.AsyncClient(verify=ssl_context) as client:
        resp = await client.post(
            url,
            json={"public_key_pem": public_key_pem, "name": name},
        )
        resp.raise_for_status()
        body = resp.json()

    ca_pem: Optional[str] = body.get("ca_cert_pem")
    if not ca_pem:
        try:
            ca_pem = await fetch_dg_ca_cert(ca_url)
        except Exception:
            ca_pem = None

    registered = ConduitIdentity(
        cert_pem=body["cert_pem"],
        key_pem=new_keypair.key_pem,
        ca_pem=ca_pem,
    )
    response = RegistrationResponse(
        id=body["id"],
        cert_pem=body["cert_pem"],
        ca_cert_pem=ca_pem,
        fingerprint=body["fingerprint"],
        name=body.get("name", name),
        registered_at=body.get("registered_at", ""),
        valid_until=body.get("valid_until"),
    )
    return registered, response


# ─── Persistence ──────────────────────────────────────────────────────────────


def save_identity(
    identity: ConduitIdentity,
    directory: Path = DEFAULT_IDENTITY_DIR,
) -> SavedIdentityPaths:
    """Save a registered identity to a directory for auto-discovery.

    Writes:
    - ``{dir}/identity.pem``     — DG-signed certificate
    - ``{dir}/identity_key.pem`` — private key (chmod 600)
    - ``{dir}/ca.pem``           — DG CA certificate (if present)

    Args:
        identity: The registered identity to persist.
        directory: Target directory (default: ``~/.conduit/``).

    Returns:
        :class:`SavedIdentityPaths` with the written file paths.
    """
    directory.mkdir(parents=True, exist_ok=True)

    cert_path = directory / "identity.pem"
    key_path = directory / "identity_key.pem"

    cert_path.write_text(identity.cert_pem)
    key_path.write_text(identity.key_pem)
    key_path.chmod(0o600)

    ca_path: Optional[Path] = None
    if identity.ca_pem:
        ca_path = directory / "ca.pem"
        ca_path.write_text(identity.ca_pem)

    return SavedIdentityPaths(cert_path=cert_path, key_path=key_path, ca_path=ca_path)


def default_identity_dir() -> Path:
    """Return the default Conduit identity directory (``~/.conduit/``)."""
    return DEFAULT_IDENTITY_DIR

File: typescript/src/registration.ts
/**
 * Substrate identity registration with the DataGrout CA.
 *
 * Flow:
 * 1. Generate an ECDSA P-256 keypair locally — private key never leaves the client.
 * 2. Send only the public key to DataGrout via {@link registerIdentity}.
 *    DG signs and returns the certificate + CA cert.
 * 3. Persist to `~/.conduit/` via {@link saveIdentity}.
 * 4. On renewal, call {@link rotateIdentity} — authenticated by mTLS, no API key needed.
 *
 * CA cert refresh
 * ---------------
 * {@link fetchDgCaCert} fetches the current DataGrout CA certificate from
 * `https://ca.datagrout.ai/ca.pem`. When the DG CA rotates, clients that call
 * {@link refreshCaCert} at startup automatically pick up the change without a
 * rebuild or re-registration.
 */

import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import * as crypto from 'crypto';

/** Canonical URL for the DataGrout CA certificate. */
export const DG_CA_URL = 'https://ca.datagrout.ai/ca.pem';

/** Default local identity directory. */
export const DEFAULT_IDENTITY_DIR = path.join(os.homedir(), '.conduit');

// ─── CA cert fetching ──────────────────────────────────────────────────────────

/**
 * Fetch the current DataGrout CA certificate from `ca.datagrout.ai`.
 *
 * Uses the system trust store for HTTPS (Cloudflare TLS) — no circularity
 * with the DG CA, which only signs client certificates.
 *
 * @param url Override the CA cert URL (default: {@link DG_CA_URL}).
 * @returns PEM-encoded CA certificate string.
 */
export async function fetchDgCaCert(url: string = DG_CA_URL): Promise<string> {
  const resp = await fetch(url, {
    headers: { Accept: 'application/x-pem-file, text/plain, */*' },
  });

  if (!resp.ok) {
    throw new Error(`CA cert fetch failed (HTTP ${resp.status}) from ${url}`);
  }

  const pem = await resp.text();

  if (!pem.includes('-----BEGIN CERTIFICATE-----')) {
    throw new Error(`Response from ${url} does not look like a PEM certificate`);
  }

  return pem;
}

/**
 * Refresh the locally-cached DG CA certificate.
 *
 * Fetches the current CA cert from `ca.datagrout.ai` and writes it to
 * `{identityDir}/ca.pem`. Call at application startup to pick up CA rotations
 * without requiring a new registration or SDK rebuild.
 *
 * @param identityDir Directory to write `ca.pem` into (default: `~/.conduit/`).
 * @param url Override the CA cert URL.
 * @returns Path to the written `ca.pem` file.
 */
export async function refreshCaCert(
  identityDir: string = DEFAULT_IDENTITY_DIR,
  url: string = DG_CA_URL
): Promise<string> {
  const pem = await fetchDgCaCert(url);
  fs.mkdirSync(identityDir, { recursive: true });
  const caPath = path.join(identityDir, 'ca.pem');
  fs.writeFileSync(caPath, pem, 'utf8');
  return caPath;
}

// ─── Types ────────────────────────────────────────────────────────────────────

/**
 * An ECDSA P-256 keypair generated locally.
 *
 * Pass to {@link registerIdentity} to exchange for a DG-signed certificate.
 * The private key never leaves the client process.
 */
export interface Keypair {
  /** PEM-encoded PKCS#8 private key (never transmitted). */
  privateKeyPem: string;
  /** PEM-encoded SPKI public key (sent to DG during registration). */
  publicKeyPem: string;
}

export interface RegistrationOptions {
  /** DataGrout substrate identity API base URL. */
  endpoint: string;
  /** Any valid DG access token for bootstrap Bearer auth. */
  authToken: string;
  /** Human-readable label for this Substrate instance. */
  name: string;
  /** CA cert URL override (default: {@link DG_CA_URL}). */
  caUrl?: string;
}

export interface RenewalOptions {
  /** DataGrout substrate identity API base URL. */
  endpoint: string;
  /** Human-readable label for the renewed identity. */
  name: string;
  /** CA cert URL override. */
  caUrl?: string;
}

export interface RegisteredIdentity {
  /** PEM-encoded DG-signed certificate. */
  certPem: string;
  /** PEM-encoded private key (never sent to DG). */
  keyPem: string;
  /** PEM-encoded DG CA certificate. */
  caPem?: string;
  /** Server-assigned identity ID. */
  id: string;
  /** Human-readable label assigned during registration. */
  name: string;
  /** SHA-256 fingerprint of the certificate. */
  fingerprint: string;
  /** ISO-8601 timestamp when the identity was registered. */
  registeredAt: string;
  /** ISO-8601 expiry date. */
  validUntil?: string;
}

// ─── Keypair generation ───────────────────────────────────────────────────────

/**
 * Generate an ECDSA P-256 keypair for Substrate identity registration.
 *
 * Returns a {@link Keypair} holding the private key (local only) and the
 * public key to send to DataGrout.  Pass the result to {@link registerIdentity}
 * to receive a DG-CA-signed certificate.
 *
 * Node.js only (`crypto.generateKeyPairSync`).
 */
export function generateKeypair(): Keypair {
  const { privateKey, publicKey } = crypto.generateKeyPairSync('ec', {
    namedCurve: 'P-256',
  });
  return {
    privateKeyPem: privateKey.export({ type: 'pkcs8', format: 'pem' }) as string,
    publicKeyPem: publicKey.export({ type: 'spki', format: 'pem' }) as string,
  };
}

// ─── Registration ─────────────────────────────────────────────────────────────

/**
 * Register a Substrate keypair with the DataGrout CA.
 *
 * Sends only the public key to DG and receives a DG-CA-signed certificate.
 * The private key in *keypair* is reused — it never leaves the client.
 *
 * @param keypair Generated by {@link generateKeypair}.
 * @param opts Registration options (endpoint, API key, name).
 */
export async function registerIdentity(
  keypair: Keypair,
  opts: RegistrationOptions
): Promise<RegisteredIdentity> {
  const { privateKeyPem, publicKeyPem } = keypair;

  const url = opts.endpoint.replace(/\/$/, '') + '/register';

  const resp = await fetch(url, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${opts.authToken}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ public_key_pem: publicKeyPem, name: opts.name }),
  });

  if (!resp.ok) {
    const body = await resp.text();
    throw new Error(`Registration failed (HTTP ${resp.status}): ${body}`);
  }

  const body = await resp.json();
  let caPem: string | undefined = body.ca_cert_pem;

  if (!caPem) {
    try {
      caPem = await fetchDgCaCert(opts.caUrl ?? DG_CA_URL);
    } catch {
      caPem = undefined;
    }
  }

  return {
    certPem: body.cert_pem,
    keyPem: privateKeyPem,
    caPem,
    id: body.id,
    name: body.name,
    fingerprint: body.fingerprint,
    registeredAt: body.registered_at ?? '',
    validUntil: body.valid_until,
  };
}

// ─── Rotation ─────────────────────────────────────────────────────────────────

export interface RotationOptions {
  /** DataGrout substrate identity API base URL. */
  endpoint: string;
  /** Human-readable label for the renewed identity. */
  name: string;
  /** CA cert URL override. */
  caUrl?: string;
  /** Current identity PEM strings, used to build the mTLS client cert. */
  currentCertPem: string;
  currentKeyPem: string;
}

/**
 * Rotate the Substrate identity by presenting the current cert over mTLS.
 *
 * Generates a new ECDSA P-256 keypair and sends the public key to the `/rotate`
 * endpoint, authenticated with the *existing* client certificate over mTLS.
 * The server returns a new DG-CA-signed certificate.
 *
 * Note: This uses Node.js `https` for mTLS — not compatible with browser environments.
 */
export async function rotateIdentity(opts: RotationOptions): Promise<RegisteredIdentity> {
  const { privateKey, publicKey } = crypto.generateKeyPairSync('ec', { namedCurve: 'P-256' });

  const publicKeyPem = publicKey.export({ type: 'spki', format: 'pem' }) as string;
  const privateKeyPem = privateKey.export({ type: 'pkcs8', format: 'pem' }) as string;

  const url = opts.endpoint.replace(/\/$/, '') + '/rotate';

  // Use Node's https module for mTLS client cert authentication.
  const https = await import('https');
  const urlModule = await import('url');
  const parsed = new urlModule.URL(url);

  const responseBody = await new Promise<string>((resolve, reject) => {
    const reqOptions = {
      hostname: parsed.hostname,
      port: parsed.port || 443,
      path: parsed.pathname,
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      cert: opts.currentCertPem,
      key: opts.currentKeyPem,
    };

    const body = JSON.stringify({ public_key_pem: publicKeyPem, name: opts.name });

    const req = https.request(reqOptions, (res) => {
      let data = '';
      res.on('data', (chunk: string) => { data += chunk; });
      res.on('end', () => {
        if (res.statusCode && res.statusCode >= 200 && res.statusCode < 300) {
          resolve(data);
        } else {
          reject(new Error(`Rotation failed (HTTP ${res.statusCode}): ${data}`));
        }
      });
    });

    req.on('error', reject);
    req.write(body);
    req.end();
  });

  const respBody = JSON.parse(responseBody);
  let caPem: string | undefined = respBody.ca_cert_pem;

  if (!caPem) {
    try {
      caPem = await fetchDgCaCert(opts.caUrl ?? DG_CA_URL);
    } catch {
      caPem = undefined;
    }
  }

  return {
    certPem: respBody.cert_pem,
    keyPem: privateKeyPem,
    caPem,
    id: respBody.id,
    name: respBody.name,
    fingerprint: respBody.fingerprint,
    registeredAt: respBody.registered_at ?? '',
    validUntil: respBody.valid_until,
  };
}

// ─── Persistence ──────────────────────────────────────────────────────────────

export interface SavedPaths {
  certPath: string;
  keyPath: string;
  caPath?: string;
}

/**
 * Save a registered identity to a directory for auto-discovery by future sessions.
 *
 * Writes:
 * - `{dir}/identity.pem`     — DG-signed certificate
 * - `{dir}/identity_key.pem` — private key (mode 0600)
 * - `{dir}/ca.pem`           — DG CA certificate (if present)
 */
export function saveIdentity(
  identity: RegisteredIdentity,
  directory: string = DEFAULT_IDENTITY_DIR
): SavedPaths {
  fs.mkdirSync(directory, { recursive: true });

  const certPath = path.join(directory, 'identity.pem');
  const keyPath = path.join(directory, 'identity_key.pem');

  fs.writeFileSync(certPath, identity.certPem, 'utf8');
  fs.writeFileSync(keyPath, identity.keyPem, { encoding: 'utf8', mode: 0o600 });

  const result: SavedPaths = { certPath, keyPath };

  if (identity.caPem) {
    const caPath = path.join(directory, 'ca.pem');
    fs.writeFileSync(caPath, identity.caPem, 'utf8');
    result.caPath = caPath;
  }

  return result;
}
